src/transaction_sender/service.zig

```diff
@@ -0,0 +1,349 @@
+/// they are confirmed, failed, timed out, or reach the max retries.
+const TransactionPool = std.AutoArrayHashMap(Signature, TransactionInfo);
+
+/// Basic send transaction service that listens for transactions on a channel.
+/// Transactions are added to the pool and retried until they are confirmed, failed,
+/// timed out, or reach the max retries.
+/// The leader schedule and current slot are loaded via RPC calls to the cluster.
+/// The leader TPU addresses are loaded from the gossip table.
+/// TODO:
+/// - Add logging
+/// - Add stats tracking
+/// - Add nonce handling
+/// - Remove RPC calls
+pub fn run(
+    allocator: std.mem.Allocator,
+    config: Config,
+    transaction_channel: *Channel(TransactionInfo),
+    gossip_table_rw: *RwMux(GossipTable),
+    exit: *AtomicBool,
+    logger: Logger,
+) !void {
+    var socket = UdpSocket.create(.ipv4, .udp) catch return error.SocketCreateFailed;
+    socket.bindToPort(config.socket.port()) catch return error.SocketBindFailed;
+    socket.setReadTimeout(socket_utils.SOCKET_TIMEOUT_US) catch return error.SocketSetTimeoutFailed;
+
+    const packet_channel = Channel(std.ArrayList(Packet)).init(allocator, 100);
+
+    const cluster_info = try LeaderInfo.init(allocator, config.cluster, gossip_table_rw);
+    var cluster_info_rw = RwMux(LeaderInfo).init(cluster_info);
+
+    const transaction_pool = TransactionPool.init(allocator);
+    var transaction_pool_rw = RwMux(TransactionPool).init(transaction_pool);
+
+    const send_socket_handle = try std.Thread.spawn(
+        .{},
+        socket_utils.sendSocket,
+        .{
+            socket,
+            packet_channel,
+            exit,
+            logger,
+        },
+    );
+
+    const receive_transactions_handle = try std.Thread.spawn(
+        .{},
+        receiveTransactionsThread,
+        .{
+            allocator,
+            transaction_channel,
+            packet_channel,
+            config,
+            &cluster_info_rw,
+            &transaction_pool_rw,
+            exit,
+            logger,
+        },
+    );
+
+    const process_transactions_handle = try std.Thread.spawn(
+        .{},
+        processTransactionsThread,
+        .{
+            allocator,
+            packet_channel,
+            config,
+            &cluster_info_rw,
+            &transaction_pool_rw,
+            exit,
+            logger,
+        },
+    );
+
+    send_socket_handle.join();
+    receive_transactions_handle.join();
+    process_transactions_handle.join();
+}
+
+fn receiveTransactionsThread(
+    allocator: std.mem.Allocator,
+    receiver: *Channel(TransactionInfo),
+    sender: *Channel(std.ArrayList(Packet)),
+    config: Config,
+    cluster_info_rw: *RwMux(LeaderInfo),
+    transaction_pool_rw: *RwMux(TransactionPool),
+    exit: *AtomicBool,
+    logger: Logger,
+) !void {
+    errdefer exit.store(true, .unordered);
+
+    var last_batch_sent = try Instant.now();
+
+    var transaction_batch = TransactionPool.init(allocator);
```

Comment:
can you add a comment or more context around the difference between the tx_batch and tx_pool

Reply:
417b8ca